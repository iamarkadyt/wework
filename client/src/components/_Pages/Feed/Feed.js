import React, { Component } from 'react'
import './Feed.scss'
import { connect } from 'react-redux'
import { compose } from 'recompose'

import Reply from '../../Reply/Reply'
import { fetchPosts, addPost } from '../../../state/actions/postsActions'
import { withEither, withAdded } from '../../../hocs/conditionalRendering'
import FBSpinner from '../../FBSpinner/FBSpinner'
import NoContent from './NoContent/NoContent'
import FeedContent from './FeedContent/FeedContent'
import EndOfFeedMessage from './EndOfFeedMessage/EndOfFeedMessage'
import QuickStats from '../../QuickStats/QuickStats'
import Discover from '../../Discover/Discover'

class Feed extends Component {
  state = {
    loadMore: false,
    isLoading: false
  }

  isEmptyFn({ endOfFeed, posts }) {
    return posts.length === 0 && !!endOfFeed
  }

  isLoadingFn({ endOfFeed, posts, isLoading }) {
    return isLoading || (!endOfFeed && posts.length === 0)
  }

  isEndOfFeedFn({ endOfFeed, isLoading }) {
    return !!endOfFeed && !isLoading
  }

  isBottom(el) {
    return el.getBoundingClientRect().bottom <= window.innerHeight
  }

  onScroll() {
    const { errors: { endOfFeed } } = this.props

    if (endOfFeed) return

    if (!this.state.isLoading) {
      if (this.isBottom(document.getElementById('Feed-mid-column'))) {
        this.setState({ loadMore: true })
      }
    }
  }

  afterFetch() {
    this.setState({ isLoading: false })
  }

  componentDidUpdate() {
    if (this.state.loadMore) {
      this.setState({ loadMore: false, isLoading: true })

      const { posts, fetchPosts } = this.props

      fetchPosts(
        posts[posts.length - 1].date,
        this.afterFetch.bind(this),
        this.afterFetch.bind(this)
      )
    }
  }

  componentDidMount() {
    const { fetchPosts } = this.props
    document.addEventListener('scroll', this.onScroll.bind(this))

    window.scrollTo(0, 0)
    fetchPosts(false)
  }

  componentWillUnmount() {
    document.removeEventListener('scroll', this.onScroll.bind(this))
  }

  render() {
    const {
      // endOfFeed is conditionally updated in state/reducers/errReducer.js
      errors: { endOfFeed },
      posts,
      addPost,
      match
    } = this.props

    const withCondRendering = compose(
      withEither(this.isEmptyFn, NoContent),
      withAdded(this.isLoadingFn, FBSpinner),
      withAdded(this.isEndOfFeedFn, EndOfFeedMessage)
    )

    const FeedContentWithCondRendering = withCondRendering(FeedContent)

    const baseUrl = match.url || ''

    return (
      <div className="Feed-container">
        <div className="Feed-lft-column">
          <QuickStats />
        </div>
        <div id="Feed-mid-column" className="Feed-mid-column">
          <Reply onSubmit={(data, callback) => addPost(data, callback)} />
          <FeedContentWithCondRendering
            isLoading={this.state.isLoading}
            endOfFeed={endOfFeed}
            posts={posts}
            baseUrl={baseUrl} />
        </div>
        <div className="Feed-rgt-column">
          <Discover />
        </div>
      </div>
    )
  }
}

export const mapStateToProps = state => ({
  posts: state.posts,
  errors: state.err
})

export { Feed }
export default connect(mapStateToProps, { fetchPosts, addPost })(Feed)
