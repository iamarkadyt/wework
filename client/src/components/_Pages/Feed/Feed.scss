.Feed-container {
    display: grid;
    grid-template-columns: 20rem 40rem 20rem;
    grid-column-gap: 1rem;
    padding-bottom: 8rem;
    .Feed-message {
        display: block;
        margin: 2.85rem 0;
        padding: .25rem 2rem;
        font-size: 1.1rem;
        background: #eee;
        color: #555;
        border-radius: .2rem;
        align-self: center;
    }
    .lds-facebook {
        align-self: center;
    }
    .Feed-mid-column {
        display: flex;
        flex-direction: column;
        align-items: stretch;
    }
}