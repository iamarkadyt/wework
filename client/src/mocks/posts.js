export const mockPost = {
    "_id": "c717a50ac29bacb2ee",
    "user": {
        "_id": "some_alphanumeric_gibberish",
        "name": "Dwayne Johnson",
        "avatar": ""
    },
    "text": "Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat.",
    "comments": [
        {
            "user": {
                "_id": "some_alphanumeric_gibberish2",
                "name": "Vin Diesel",
                "avatar": ""
            },
            "text": "Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.",
            "date": "1535133201"
        },
        {
            "user": {
                "_id": "some_alphanumeric_gibberish2",
                "name": "Vin Diesel",
                "avatar": ""
            },
            "text": "Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.",
            "date": "1535133201"
        },
        {
            "user": {
                "_id": "some_alphanumeric_gibberish2",
                "name": "Vin Diesel",
                "avatar": ""
            },
            "text": "Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.",
            "date": "1535133201"
        }
    ],
    "likes": [
        {
            "_id": "19825918265102965",
            "user": "aoeu9a8o10s392h0"
        },
        {
            "_id": "19825918265102966",
            "user": "aoeu9a8o10s392hx0"
        },
        {
            "_id": "19825918265102967",
            "user": "aoeu9a8o10s392120"
        }
    ],
    "date": "1535133024"
}

export default new Array(10).fill(mockPost, 0, 10)
