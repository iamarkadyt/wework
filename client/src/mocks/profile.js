export default {
    "skills": [
        "Assembler",
        "Go",
        "C#",
        "C++"
    ],
    "social": [],
    "date": "2018-08-23T20:25:12.263Z",
    "_id": "5b7b42ba57a5806b896e8f0d",
    "user": {
        "_id": "5b75f023f1957052e52106f5",
        "name": "John Doe"
    },
    "__v": 0,
    "handle": "techguy_handle",
    "status": "Actively applying",
    "title": "Senior Developer",
    "company": "Itron",
    "experience": [
        {
            "current": false,
            "_id": "5b7ca6cf63353349cf03756c",
            "title": "SWE",
            "company": "Revature",
            "from": "2009-03-31T23:00:00.000Z",
            "to": "2025-08-01T01:13:50.111Z"
        }
    ],
    "education": [
        {
            "current": true,
            "_id": "5b7cab633345cc4c066c0540",
            "school": "AC Bootcamp",
            "degree": "Nano CS Degree",
            "from": "2015-08-02T18:33:20.000Z"
        }
    ],
    "bio": "My name is [techguyinfo]. I am a senior developer currently working at Microsoft. I specialize in AI and Cloud computing.",
    "githubusername": "arkadyt",
    "location": "Seattle, WA",
    "website": "techguy.info"
}