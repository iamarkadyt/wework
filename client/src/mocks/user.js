export default {
  id: "98ac61abed72ee8d12"
}
