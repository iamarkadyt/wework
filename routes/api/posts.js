const express = require('express');
const router = express.Router();

// Can setup middleware like this:
// router.use(functions);

// @route   GET api/posts/test
// @desc    Tests posts route.
// @access  Public
router.get('/test', (req, res) => {
    res.send('OK');
});

module.exports = router;