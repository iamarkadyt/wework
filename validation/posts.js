const Yup = require('yup')

module.exports = profile => new Promise((resolve, reject) => {
    const schema = Yup.object().shape({
        text: Yup.string()
            .min(10, "Minimum of 10 characters is required")
            .max(300, "Maximum of 300 characters is allowed")
            .required("Please enter some text here")
    })

    schema.validate(profile, { abortEarly: false, strict: true })
        .then(profile => resolve(profile))
        .catch(err => {
            const errors = {}
            for (let i of err.inner)
                errors[i.path] = i.message
            reject(errors)
        })
})