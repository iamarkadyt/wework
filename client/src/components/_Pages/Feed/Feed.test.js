import React from 'react'
import { shallow } from 'enzyme'
import { Feed, mapStateToProps } from './Feed'
import cloneDeep from 'lodash.clonedeep'
import { mockPost } from '../../../mocks/posts'

let mountedComponent, props

const getMockProps = () => {
  return cloneDeep({
    errors: {},
    match: { url: '' },
    fetchPosts: jest.fn(),
    posts: [mockPost]
  })
}

const comp = () => {
  if (!mountedComponent) {
    mountedComponent = shallow(<Feed {...props} />)
  }
  return mountedComponent
}

describe('Feed', () => {
  beforeEach(() => {
    mountedComponent = undefined
    props = getMockProps()

    global.scrollTo = jest.fn()

    const listeners = {}
    global.document.addEventListener = jest.fn((name, fn) => listeners[name] = fn)
    global.document.removeEventListener = jest.fn((name, fn) => delete listeners[name])
  })

  describe('props', () => {
    describe('passing', () => {
      describe('Reply component', () => {
        it('receives correct onSubmit prop', () => {
          props.addPost = jest.fn()
          comp().find('Connect(Reply)').simulate('submit', 1, 2)
          expect(props.addPost).toHaveBeenCalledWith(1, 2)
        })
      })
    })
  })

  describe('functions', () => {
    describe('componentWillUnmount', () => {
      it('removes scroll event listener', () => {
        comp().unmount()
        // TODO better function matching, expect.extend(...)
        expect(global.document.removeEventListener).toHaveBeenCalledWith('scroll', expect.any(Function))
      })
    })

    describe('componentDidMount', () => {
      it('calls fetchPosts with...', () => {
        comp()
        expect(props.fetchPosts).toHaveBeenCalledWith(false)
      })

      it('calls fetchPosts once', () => {
        comp()
        expect(props.fetchPosts).toHaveBeenCalledTimes(1)
      })

      it('calls scrollTo with...', () => {
        comp()
        expect(global.scrollTo).toHaveBeenCalledWith(0, 0)
      })

      it('calls scrollTo once', () => {
        comp()
        expect(global.scrollTo).toHaveBeenCalledTimes(1)
      })

      it('calls addEventListener with...', () => {
        comp()
        // TODO better function matching, expect.extend(...)
        expect(global.document.addEventListener).toHaveBeenCalledWith('scroll', expect.any(Function))
      })

      it('calls addEventListener once', () => {
        comp()
        expect(global.document.addEventListener).toHaveBeenCalledTimes(1)
      })
    })

    describe('componentDidUpdate', () => {
      describe('if state.loadMore is true', () => {
        it('updates state as expected', () => {
          comp().setState({ loadMore: true })
          expect(comp().state()).toEqual({ loadMore: false, isLoading: true })
        })

        it('calls fetchPosts once', () => {
          comp()
          props.fetchPosts.mockClear()
          comp().setState({ loadMore: true })
          expect(props.fetchPosts).toHaveBeenCalledTimes(1)
        })

        it('calls fetchPosts with...', () => {
          comp().setState({ loadMore: true })
          const { posts, fetchPosts } = props

          // TODO test which functions exactly
          expect(fetchPosts).toHaveBeenCalledWith(
            posts[posts.length - 1].date, 
            expect.any(Function), 
            expect.any(Function)
          )
        })
      })

      describe('if state.loadMore is false', () => {
        // TODO find a better way to assert that nothing happens

        it('does not update state', () => {
          const prevState = { ...comp().state() }
          comp().setState({ loadMore: false })
          expect(comp().state()).toStrictEqual(prevState)
        })

        it('does not call fetchPosts', () => {
          comp()
          props.fetchPosts.mockClear()
          comp().setState({ loadMore: false })
          expect(props.fetchPosts).not.toHaveBeenCalled()
        })
      })
    })

    describe('afterFetch', () => {
      it('calls setState', () => {
        const fn = jest.spyOn(Feed.prototype, 'setState')
        comp()
        fn.mockClear()
        comp().instance().afterFetch()
        expect(fn).toHaveBeenCalledWith({ isLoading: false })
      })
    })

    describe('onScroll', () => {
      it('if errors.endOfFeed is defined, does nothing', () => {
        props.errors.endOfFeed = 'some error, sir!'
        const setState = jest.spyOn(Feed.prototype, 'setState')

        comp()
        setState.mockClear()

        comp().instance().onScroll()
        expect(setState).not.toHaveBeenCalled()
      })

      it('if state.isLoading is true, does nothing', () => {
        delete props.errors.endOfFeed
        const setState = jest.spyOn(Feed.prototype, 'setState')

        comp().setState({ isLoading: true })
        setState.mockClear()

        comp().instance().onScroll()
        expect(setState).not.toHaveBeenCalled()
      })

      it('if not page bottom, does nothing', () => {
        delete props.errors.endOfFeed
        const setState = jest.spyOn(Feed.prototype, 'setState')
        const isBottom = jest.spyOn(Feed.prototype, 'isBottom')
        isBottom.mockImplementationOnce(() => false)

        comp().setState({ isLoading: false })
        setState.mockClear()

        comp().instance().onScroll()
        expect(setState).not.toHaveBeenCalled()
      })

      it('if all above is not true, updates the state', () => {
        delete props.errors.endOfFeed
        const setState = jest.spyOn(Feed.prototype, 'setState')
        const isBottom = jest.spyOn(Feed.prototype, 'isBottom')
        isBottom.mockImplementationOnce(() => true)

        comp().setState({ isLoading: false })
        setState.mockClear()

        comp().instance().onScroll()
        expect(setState).toHaveBeenCalled()
      })
    })

    describe('isBottom', () => {
      it('calls getBoundingClientRect', () => {
        const el = {
          getBoundingClientRect: jest.fn(() => ({ bottom: 0 }))
        }
        comp().instance().isBottom(el)
        expect(el.getBoundingClientRect).toHaveBeenCalled()
      })

      it('returns true page bottom is reached', () => {
        global.window.innerHeight = 100
        const el = {
          getBoundingClientRect: jest.fn(() => ({ bottom: 100 }))
        }
        expect(comp().instance().isBottom(el)).toBe(true)
      })

      it('returns false if page bottom is not reached', () => {
        global.window.innerHeight = 20
        const el = {
          getBoundingClientRect: jest.fn(() => ({ bottom: 100 }))
        }
        expect(comp().instance().isBottom(el)).toBe(false)
      })
    })

    describe('isEndOfFeedFn', () => {
      it('returns true if endOfFeed is defined and isLoading is false', () => {
        const customProps = {
          endOfFeed: 'some error',
          isLoading: false
        }
        expect(comp().instance().isEndOfFeedFn(customProps)).toBe(true)
      })

      it('returns false if endOfFeed is not defined but isLoading is false', () => {
        const customProps = {
          isLoading: false
        }
        expect(comp().instance().isEndOfFeedFn(customProps)).toBe(false)
      })

      it('returns false if endOfFeed is defined but isLoading is true', () => {
        const customProps = {
          endOfFeed: 'some error',
          isLoading: true
        }
        expect(comp().instance().isEndOfFeedFn(customProps)).toBe(false)
      })

      it('returns false if endOfFeed is defined and isLoading is true', () => {
        const customProps = {
          endOfFeed: 'some error',
          isLoading: true
        }
        expect(comp().instance().isEndOfFeedFn(customProps)).toBe(false)
      })
    })

    describe('isLoadingFn', () => {
      it('returns true if isLoading is true', () => {
        const customProps = {
          isLoading: true
        }
        expect(comp().instance().isLoadingFn(customProps)).toBe(true)
      })

      it('returns true if isLoading is false but endOfFeed is not defined and no posts were loaded', () => {
        const customProps = {
          isLoading: false,
          posts: []
        }
        expect(comp().instance().isLoadingFn(customProps)).toBe(true)
      })

      it('returns true if isLoading is true and endOfFeed is not defined and no posts were loaded', () => {
        const customProps = {
          isLoading: true,
          posts: []
        }
        expect(comp().instance().isLoadingFn(customProps)).toBe(true)
      })

      it('returns false if isLoading is false and endOfFeed is defined but no posts were loaded', () => {
        const customProps = {
          isLoading: false,
          posts: [],
          endOfFeed: 'error!'
        }
        expect(comp().instance().isLoadingFn(customProps)).toBe(false)
      })

      it('returns false if isLoading is false and endOfFeed is not defined but some posts were loaded', () => {
        const customProps = {
          isLoading: false,
          posts: [ {}, {}, {} ],
        }
        expect(comp().instance().isLoadingFn(customProps)).toBe(false)
      })

      it('returns false if isLoading is false, endOfFeed is defined and some posts were loaded', () => {
        const customProps = {
          isLoading: false,
          endOfFeed: 'error!',
          posts: [ {}, {}, {} ],
        }
        expect(comp().instance().isLoadingFn(customProps)).toBe(false)
      })
    })

    describe('isEmptyFn', () => {
      it('returns true if no posts were loaded and endOfFeed is defined', () => {
        const customProps = {
          endOfFeed: 'defined!',
          posts: []
        }
        expect(comp().instance().isEmptyFn(customProps)).toBe(true)
      })

      it('returns false if some posts were loaded and endOfFeed is defined', () => {
        const customProps = {
          endOfFeed: 'defined!',
          posts: [ {} ]
        }
        expect(comp().instance().isEmptyFn(customProps)).toBe(false)
      })

      it('returns false if no posts were loaded but endOfFeed is undefined', () => {
        const customProps = {
          posts: []
        }
        expect(comp().instance().isEmptyFn(customProps)).toBe(false)
      })

      it('returns false if some posts were loaded and endOfFeed is undefined', () => {
        const customProps = {
          posts: [ {} ],
          endOfFeed: 'some err'
        }
        expect(comp().instance().isEmptyFn(customProps)).toBe(false)
      })
    })
  })

  describe('mapStateToProps', () => {
    it('returs expected object', () => {
      const state = {
        posts: [],
        err: {}
      }

      const expectedObj = {
        posts: state.posts,
        errors: state.err
      }

      expect(mapStateToProps(state)).toStrictEqual(expectedObj)
    })
  })

  it('matches snapshot', () => {
    expect(comp()).toMatchSnapshot()
  })
})
