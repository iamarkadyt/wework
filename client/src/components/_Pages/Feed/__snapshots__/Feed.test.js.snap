// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Feed matches snapshot 1`] = `
<div
  className="Feed-container"
>
  <div
    className="Feed-lft-column"
  >
    <withRouter(Connect(QuickStats)) />
  </div>
  <div
    className="Feed-mid-column"
    id="Feed-mid-column"
  >
    <Connect(Reply)
      onSubmit={[Function]}
    />
    <Component
      baseUrl=""
      isLoading={false}
      posts={
        Array [
          Object {
            "__v": 0,
            "_id": "5b9f4728bc40273febf52706",
            "comments": Array [
              Object {
                "_id": "5ba1dd77809c892f580240ef",
                "date": "2018-09-19T05:24:07.399Z",
                "text": "Hahahahahahahaaa!",
                "user": Object {
                  "_id": "5b75f023f1957052e52106f5",
                  "avatar": "//www.gravatar.com/avatar/254dbff8aa12a023b9f1052ad80b1831?s=200&r=pg&d=mm",
                  "name": "John Grapard",
                },
              },
              Object {
                "_id": "5ba1de21f0b9f73054396abf",
                "date": "2018-09-19T05:26:57.632Z",
                "text": "Nuka proverim esli oshibka ostalas...",
                "user": Object {
                  "_id": "5b75f023f1957052e52106f5",
                  "avatar": "//www.gravatar.com/avatar/254dbff8aa12a023b9f1052ad80b1831?s=200&r=pg&d=mm",
                  "name": "John Grapard",
                },
              },
            ],
            "date": "2018-09-17T06:18:16.992Z",
            "likes": Array [
              Object {
                "_id": "5ba1dd61809c892f580240ed",
                "user": "5b75f023f1957052e52106f5",
              },
              Object {
                "_id": "5ba80533058247301e362bf9",
                "user": "5ba1acfc13b542138b29af68",
              },
              Object {
                "_id": "5ba847153c823e00161763b5",
                "user": "5ba83562ed730446b54def34",
              },
            ],
            "text": "Tas nisl commlor sit amet, consec!",
            "user": Object {
              "_id": "5b9eedd621194c102c5280e8",
              "avatar": "//www.gravatar.com/avatar/e13743a7f1db7f4246badd6fd6ff54ff?s=200&r=pg&d=mm",
              "name": "John Doe",
            },
          },
        ]
      }
    />
  </div>
  <div
    className="Feed-rgt-column"
  >
    <Connect(Discover) />
  </div>
</div>
`;
